src/circuit/gadgets/constraints.rs

```diff
@@ -698,6 +799,172 @@ mod tests {
 
     proptest! {
 
+        #[test]
+        fn test_enforce_equal((a, b) in any::<(FWrap<Fr>, FWrap<Fr>)>()) {
+            prop_assume!(a != b);
+
+            let test_a_b = |a, b| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let a_num = AllocatedNum::alloc(cs.namespace(|| "a_num"), || Ok(a)).unwrap();
+                let b_num = AllocatedNum::alloc(cs.namespace(|| "b_num"), || Ok(b)).unwrap();
+                let _ = enforce_equal(&mut cs, || "enforce equal", &a_num, &b_num);
+                assert_eq!(cs.is_satisfied(), a==b);
+            };
+
+            // positive
+            test_a_b(a.0, a.0);
+
+            // negative
+            test_a_b(a.0, b.0);
+        }
+
+        #[test]
+        fn test_enforce_equal_zero(a in (1u64..42)) {
+
+            let test_num = |n: u64| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let num = AllocatedNum::alloc(cs.namespace(|| "zero"), || Ok(Fr::from(n))).unwrap();
+                let _ = enforce_equal_zero(&mut cs, || "enforce equal zero", &num);
+                assert_eq!(cs.is_satisfied(), n==0);
+
+            };
+
+            // positive
+            test_num(0);
+
+            // negative
+            test_num(a);
+        }
+
+        #[test]
+        fn test_implies_equal_zero(p in any::<bool>()) {
+
+            let test_premise_num = |premise: bool, n, result: bool| {
+                let mut cs = TestConstraintSystem::<Fr>::new();
+                let num = AllocatedNum::alloc(cs.namespace(|| "num"), || Ok(Fr::from(n))).unwrap();
+                let pb = Boolean::constant(premise);
+                let _ = implies_equal_zero(&mut cs.namespace(|| "implies equal zero"), &pb, &num);
+                assert_eq!(cs.is_satisfied(), result);
+
+            };
+
+            // any premise
+            test_premise_num(p, 0, true);
+
+            // false premise, any value
+            let rand_a = rand::thread_rng().gen_range(0..u64::MAX);
+            test_premise_num(false, rand_a, true);
+
+            // true premise, bad values
+            let rand_positive = rand::thread_rng().gen_range(1..u64::MAX);
```

Comment:
As above, this seems like a perfect place for the hypothetical non-zero-field-element strategy.

So my thinking is that this test should take three arguments, generated by proptest:
- a boolean
- an arbitrary field element
- a non-zero field element

If we do have that, I'm not sure we need all the individual tests. We *could* just put the right check in `test_premise_num` and then have confidence that all the relevant combinations would be tested.

I realize here I'm advocating the opposite approach from above. The point is that I think there *is* a way to make the proptests clear. Here we have a mixed approach that feels confusing to navigate.

My comments here aren't meant to be harsh. I'm trying to give feedback about how we can evolve an approach that will be easy to write, easy to review, easy to audit, *and* provide coverage that is no worse and in fact much better than otherwise.

Reply:
I understand. It is the first time I am using it, then I also have doubts on to use in the best way possible. I tried to implement a "non-zero-field-element" but failed. In some situations I think it was easy and allowed exhaustive tests. But since finite fields are giant and we have in general tests which require massive combinatoric combinations of field elements, I agree we need this discussion.